clean-modular-code/python-pep-8.md

```diff
@@ -14,237 +14,279 @@ kernelspec:
++++ {"editable": true, "slideshow": {"slide_type": ""}}
 
-* **CamelCase:** Every word is capitalized so they visually stand out: `CapitalizedWords`. This is sometimes also referred to as CapWords or StudlyCaps.
+Notice that by adding docstrings, spacing to the the code.... 
 
-  * Note: When using acronyms in CamelCase, capitalize all the letters of the acronym. Thus HTTPServerError is better than HttpServerError.
+```{code-cell} ipython3
+---
+editable: true
+slideshow:
+  slide_type: ''
+---
+# Built-in libraries are imported first
+from datetime import datetime
+import pandas as pd
+
+# Function to classify precipitation levels
+def classify_precipitation(precip_list):
+    """Classify average monthly precipitation into low, medium, or high.
+
+    Parameters
+    ----------
+    precip_list : list of float
+        A list of monthly precipitation values (in mm).
+
+    Returns
+    -------
+    str
+        The precipitation category: 'Low', 'Medium', or 'High'.
+    """
+    avg_precip = pd.Series(precip_list).mean()
+
+    # Define the precipitation ranges
+    if avg_precip < 100:
+        return 'Low'
+    elif 100 <= avg_precip <= 150:
+        return 'Medium'
+    else:
+        return 'High'
+
+
+# Sample data for precipitation values (in mm) for different locations
+data = {
+    'location': ['Station1', 'Station2', 'Station3', 'Station4'],
+    'year': [2021, 2021, 2021, 2021],
+    'monthly_precipitation': [
+        [50.0, 70.0, 90.0, 80.0],  # Station1
+        [100.0, 110.0, 120.0, 130.0],   # Station2
+        [150.0, 160.0, 170.0, 180.0],   # Station3
+        [200.0, 210.0, 220.0, 230.0]    # Station4
+    ],
+    'start_date': ["2021-01-01", "2021-01-01", "2021-01-01", "2021-01-01"]
+}
+
+df = pd.DataFrame(data)
+df['start_date'] = pd.to_datetime(df['start_date'])
+
+# Classify precipitation levels based on ranges
+df['precipitation_category'] = df['monthly_precipitation'].apply(classify_precipitation)
+
+df
+```
 
-* **mixedCase:** (differs from CapitalizedWords by initial lowercase character!)
++++ {"editable": true, "slideshow": {"slide_type": ""}}
 
-* **Capitalized_Words_With_Underscores:** This approach is not recommended. Use one convention and stick with it.
+## How to Apply PEP 8 Code Style Standards
 
-### Name Variables Using snake_case And All Lower Case
+It may seem overwhelming to remember all the PEP 8 rules, but tools called **code formatters** can automatically apply these standards for you.
 
-In general, it is recommended that you keep naming conventions standard in your code. We suggest a convention that uses **snake_case** and all lowercase letters in your code for variable and function names.
+### For `.py` Files
 
-```
-variable_one
-variable_two
-```
+Use popular tools like **Black** or **Ruff**:
 
-### Name Classes Using CamelCase or CapsCase
+- **[Black](https://black.readthedocs.io/en/stable/)**: Automatically reformats code according to PEP 8.
```

Comment:
As a co-maintainer of Black, the original intent was to apply rules to minimize the size of diffs and to make code review easier. 

@sneakers-the-rat I would be interested when black and ruff contradict each other.

Reply:
That's the main one I have experienced! Mostly when writing `@override`s, black will want to put the ellipsis on the same line and ruff will want to put it on the following line, its not a big deal I just have to remember to run black after ruff which is what I think you're supposed to do anyway?

Reply:
So, I've stopped using Ruff on some projects and have migrated to Ruff for code formatting, too, for ease of tooling. But i still use black as a code formatter in vscode. It's so interesting that there are so many different options when thinking about linting and code formatting tools. I say this because only a few years ago i was using autopep8 for software and the autopep8 tool for jupyter in teaching.